File: nest-sso-backend/src/auth/auth.controller.ts
import {
  BadRequestException,
  Controller,
  Get,
  HttpCode,
  HttpException,
  HttpStatus,
  Post,
  Redirect,
  Req,
  Res,
  UnauthorizedException,
} from '@nestjs/common';
import { AuthService } from './auth.service';
import { ConfigService } from '@nestjs/config';
import { Request, Response } from 'express';

@Controller('auth')
export class AuthController {
  constructor(
    private readonly authService: AuthService,
    private readonly configService: ConfigService,
  ) {}

  @Get('login')
  @Redirect()
  async login(@Res({ passthrough: true }) res: Response) {
    const data = await this.authService.generatePkce();
    const { redirectUrl, codeVerifier } = data;
    res.cookie('pkce_verifier', codeVerifier, {
      httpOnly: true,
      secure: true,
      sameSite: 'none',
      path: '/',
    });
    return { url: redirectUrl };
  }

  @Get('callback')
  async callback(@Req() req: Request, @Res() res: Response) {
    const { code, state } = req.query as { code: string; state: string };
    if (!code || !state) {
      throw new BadRequestException('Missing code or state in callback');
    }

    const codeVerifier = req.cookies['pkce_verifier'];
    if (!codeVerifier) {
      throw new UnauthorizedException('PKCE code verifier missing');
    }
    const tokens = await this.authService.handleCallback(code, codeVerifier);

    res.clearCookie('pkce_verifier', {
      httpOnly: true,
      secure: true,
      sameSite: 'none',
      path: '/',
    });
    res.cookie('refresh_token', tokens.refreshToken, {
      httpOnly: true,
      secure: true,
      sameSite: 'none',
      maxAge: tokens.refreshExpiresIn * 1000,
      domain: '.quijadajosed.duckdns.org',
      path: '/',
    });
    const angularCallbackUrl = `${this.configService.get<string>('SPA_BASE_URL')}/auth-callback`;
    res.redirect(angularCallbackUrl);
  }

  @Post('refresh')
  @HttpCode(200)
  async refresh(@Req() req: Request, @Res() res: Response) {
    const refreshToken = req.cookies['refresh_token'];
    if (!refreshToken) {
      throw new UnauthorizedException('No refresh token provided');
    }

    try {
      const tokens = await this.authService.refreshToken(refreshToken);
      res.cookie('refresh_token', tokens.refreshToken, {
        httpOnly: true,
        secure: true,
        sameSite: 'none',
        maxAge: tokens.refreshExpiresIn * 1000,
        domain: '.quijadajosed.duckdns.org',
        path: '/',
      });
      return res.json({
        access_token: tokens.accessToken,
        expires_in: tokens.accessExpiresIn,
        token_type: 'Bearer',
      });
    } catch (error) {
      if (error instanceof UnauthorizedException) {
        throw error;
      }

      throw new HttpException(
        'Failed to refresh token',
        HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }

  @Post('logout')
  @HttpCode(204)
  async logout(@Req() req: Request, @Res({ passthrough: true }) res: Response) {
    const refresh_token = req.cookies['refresh_token'];
    if (!refresh_token) {
      throw new UnauthorizedException('Missing refresh_token for logout');
    }

    await this.authService.logout(refresh_token);
    res.clearCookie('refresh_token', {
      httpOnly: true,
      secure: true,
      sameSite: 'none',
      domain: '.quijadajosed.duckdns.org',
      path: '/',
    });
  }
}

File: spring-sso-backend/src/main/java/com/ejemplo/backend/controller/AuthController.java
package com.ejemplo.backend.controller;

import com.ejemplo.backend.service.AuthService;
import com.ejemplo.backend.service.AuthService.PkceResult;
import com.ejemplo.backend.service.AuthService.TokenResult;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

@RestController
@RequestMapping("/auth")
public class AuthController {

    private final AuthService authService;

    @Value("${spa.base-url}")
    private String spaBaseUrl;

    public AuthController(AuthService authService) {
        this.authService = authService;
    }

    @GetMapping("/login")
    public ResponseEntity<Void> login(HttpServletResponse response) {
        PkceResult pkce = authService.generatePkce();

        ResponseCookie cookie = ResponseCookie.from("pkce_verifier", pkce.codeVerifier())
                .httpOnly(true)
                .secure(true)
                .path("/")
                .maxAge(300)
                .sameSite("None")
                .build();
        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());

        return ResponseEntity.status(HttpStatus.FOUND)
                .header(HttpHeaders.LOCATION, pkce.redirectUrl())
                .build();
    }

    @GetMapping("/callback")
    public void callback(
            @RequestParam String code,
            @RequestParam String state,
            HttpServletRequest request,
            HttpServletResponse response
    ) throws Exception {
        String codeVerifier = null;
        for (Cookie c : request.getCookies()) {
            if ("pkce_verifier".equals(c.getName())) {
                codeVerifier = c.getValue();
            }
        }
        if (codeVerifier == null) {
            response.sendError(HttpStatus.UNAUTHORIZED.value(), "PKCE verifier missing");
            return;
        }

        TokenResult tokens = authService.handleCallback(code, codeVerifier);

        ResponseCookie clearPkce = ResponseCookie.from("pkce_verifier", "")
                .httpOnly(true).secure(true).path("/")
                .maxAge(0).sameSite("None").build();
        response.addHeader(HttpHeaders.SET_COOKIE, clearPkce.toString());

        ResponseCookie refresh = ResponseCookie.from("refresh_token", tokens.refreshToken())
                .httpOnly(true).secure(true).path("/")
                .maxAge(tokens.refreshExpiresIn()).sameSite("None").build();
        response.addHeader(HttpHeaders.SET_COOKIE, refresh.toString());

        response.sendRedirect(spaBaseUrl + "/auth-callback");
    }

    @PostMapping("/refresh")
    @ResponseStatus(HttpStatus.OK)
    public ResponseEntity<Object> refresh(HttpServletRequest request, HttpServletResponse response) {
        String refreshToken = null;
        for (Cookie c : request.getCookies()) {
            if ("refresh_token".equals(c.getName())) {
                refreshToken = c.getValue();
            }
        }
        if (refreshToken == null) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
        }

        TokenResult tokens = authService.refreshToken(refreshToken);

        ResponseCookie refresh = ResponseCookie.from("refresh_token", tokens.refreshToken())
                .httpOnly(true).secure(true).path("/")
                .maxAge(tokens.refreshExpiresIn())
                .sameSite("None").build();
        response.addHeader(HttpHeaders.SET_COOKIE, refresh.toString());

        return ResponseEntity.ok(Map.of(
                "access_token", tokens.accessToken(),
                "expires_in", tokens.accessExpiresIn(),
                "token_type", "Bearer"
        ));
    }

    @PostMapping("/logout")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    public void logout(HttpServletRequest request, HttpServletResponse response) {
        String refreshToken = null;
        for (Cookie c : request.getCookies()) {
            if ("refresh_token".equals(c.getName())) {
                refreshToken = c.getValue();
            }
        }
        if (refreshToken != null) {
            authService.logout(refreshToken);
        }
        ResponseCookie clear = ResponseCookie.from("refresh_token", "")
                .httpOnly(true).secure(true).path("/")
                .maxAge(0).sameSite("None").build();
        response.addHeader(HttpHeaders.SET_COOKIE, clear.toString());
    }
}
